queryserver-orchestrator/src/main/java/org/apache/phoenix/queryserver/orchestrator/QueryServerCanaryOrchestrator.java

```diff
@@ -0,0 +1,117 @@
+
+		parser.addArgument("--zkpath", "-zkp").type(String.class).nargs("?").setDefault(PATH)
+				.help("ZKNode path default: " + PATH);
+
+		parser.addArgument("--hostname", "-hn").type(String.class).nargs("?").help("Hostname on "
+				+ "which PQS is running.");
+
+		parser.addArgument("--port", "-p").type(String.class).nargs("?").help("Port on which PQS " +
+				"" + "" + "" + "" + "" + "" + "" + "is running.");
+
+		parser.addArgument("--constring", "-cs").type(String.class).nargs("?").help("Pass an " +
+				"explicit " + "connection String to connect to PQS. default: null");
+
+		parser.addArgument("--timeout", "-t").type(String.class).nargs("?").setDefault("120")
+				.help("Maximum time for " + "which the app should run before returning error. " +
+						"default: 120 sec");
+
+		parser.addArgument("--testtable", "-tt").type(String.class).nargs("?").setDefault
+				("PQSTEST").help("Custom " + "name for the test table. default: PQSTEST");
+
+		parser.addArgument("--logsinkclass", "-lsc").type(String.class).nargs("?").setDefault
+				("org.apache.phoenix" + ".queryserver.tool.QueryServerCanaryTool$StdOutSink")
+				.help("Path to a Custom implementation for log " + "sink class. default: stdout");
+
+		parser.addArgument("--zkurl", "-zk").type(String.class).help("URL for Zookeeper");
+
+		parser.addArgument("--interval", "-in").type(String.class).nargs("?").setDefault("900")
+				.help("Time interval " + "between 2 consecutive test suite runs");
+
+		Namespace res = null;
+		try {
+			res = parser.parseArgs(args);
+		} catch (ArgumentParserException e) {
+			parser.handleError(e);
+		}
+		return res;
+	}
+
+	public static Map<String, String> getArgs(Namespace cArgs) {
+
+		Map<String, String> params = new HashMap<>();
+
+		for (Map.Entry<String, Object> entry : cArgs.getAttrs().entrySet()) {
+			params.put(entry.getKey(), (String) entry.getValue());
+		}
+		return params;
+	}
+
+	public static void main(String[] args) {
+
+		logger.info("Starting PQS Canary Orchestrator...");
+		ToolWrapper tool = new ToolWrapper();
+		try {
+
+			Namespace cArgs = parseArgs(args);
+			if (cArgs == null) {
+				throw new RuntimeException("Argument parsing failed");
+			}
+
+			Map<String, String> params = getArgs(cArgs);
+			PATH = params.get("zkpath");
+
+			CuratorFramework client = CuratorFrameworkFactory.newClient(params.get("zkurl"), new
+					BoundedExponentialBackoffRetry(1000, 60000, 50000));
+
+
+			TestExecutorClient executorClient = new TestExecutorClient(client, PATH, params, tool);
+			client.start();
+			executorClient.start();
+
+			// A shutdown hook to cleanly terminate the application
+			Runtime.getRuntime().addShutdownHook(new Thread() {
+				public void run() {
+					logger.info("Shutdown Hook for PQS Canary Orchestrator running...");
+					CloseableUtils.closeQuietly(executorClient);
+					CloseableUtils.closeQuietly(client);
+					logger.info("Closed Curator Client");
+				}
+			});
+
+			while (true) {
```

Comment:
This causes 100% CPU utilization and should not be used. Ideally this process should spawn off a single new thread and call the join method here. https://stackoverflow.com/questions/14763434/how-to-avoid-100-cpu-utilization-without-removing-whiletrue

Reply:
done

Reply:
Since the dependent thread is a curator client which depends on the framework to call takeLeadership() method. This process is oblivious to the thread that it might depend on, hence the join on testExecutorClient will see the thread execution is completed and let main thread exit. 

Any suggestion on how do we handle this here?

Reply:
You want to implement some logic in `TestExecutorClient` that this caller can block on.

For `Thread`, this would be doing something like calling `join()` on a thread you started which you want to wait to complete. Another way to do this would be to make a boolean method `isDone()`on `TestExecutorClient` which this class can call periodically, e.g.
```
while (!client.isDone()) {
  Thread.sleep(1, TimeUnit.SECOND);
}
```

Reply:
Thank you, @joshelser. 
join didnt help this scenario so used other approach.